zipline-gradle-plugin/src/main/kotlin/app/cash/zipline/gradle/ZiplineDownloader.kt

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (C) 2021 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package app.cash.zipline.gradle
+
+import app.cash.zipline.loader.DriverFactory
+import app.cash.zipline.loader.OkHttpZiplineHttpClient
+import app.cash.zipline.loader.ZiplineLoader
+import java.io.File
+import java.time.Clock
+import java.util.concurrent.Executors
+import kotlinx.coroutines.asCoroutineDispatcher
+import kotlinx.coroutines.runBlocking
+import okhttp3.HttpUrl.Companion.toHttpUrl
+import okhttp3.OkHttpClient
+import okio.FileSystem
+import okio.Path.Companion.toOkioPath
+import okio.Path.Companion.toPath
+
+class ZiplineDownloader {
+  private val executorService = Executors.newSingleThreadExecutor { Thread(it, "Zipline") }
+  private val dispatcher = executorService.asCoroutineDispatcher()
+  private val client = OkHttpClient()
+
+  // Dummy base url
+  private val baseUrl = "http://10.0.2.2:8080/".toHttpUrl()
+
+  // Run SQLite in memory
+  private val driver = DriverFactory().createDriver()
```

Comment:
The need to do this makes me feel like maybe the layering here is slightly off or perhaps we're giving too much responsibility to a single type. I don't know the implementation well enough to give a concrete suggestions.

Right now this type knows about the local cache, the embedded cache, and the HTTP source. Am I right in thinking that we can layer these into an interceptor-like pattern? Where we have something like an `HttpModuleLoader`, an `FsCacheModuleLoader`, and an `EmbeddedModuleLoader` and they're a hierarchy which delegates to the next on a miss?

That would remove knowledge of the HTTP client, the filesystem, and the database from ZiplineLoader and have it take a single `ModuleLoader`.

It probably also eliminates the need for the strategy classes because the strategy just becomes how you layer the various loaders while also de-duplicating them since both strategies currently have the same logic for downloading.

Reply:
Good call, I like how the interceptor pattern would better fit here and prevent having to plumb all possible dependencies into the ZiplineLoader constructor which as you've seen has made testing more difficult.

Reply:
Can be for a follow-up PR rather than keeping this open